core/src/main/java/jenkins/util/AtmostOneTaskExecutor.java

```diff
@@ -65,10 +72,10 @@ public AtmostOneTaskExecutor(ExecutorService base, Callable<V> task) {
     }
 
     public AtmostOneTaskExecutor(Callable<V> task) {
-        this(new AtmostOneThreadExecutor(new NamingThreadFactory(
+        this(new ImpersonatingExecutorService(new AtmostOneThreadExecutor(new NamingThreadFactory(
                         new DaemonThreadFactory(),
                         String.format("AtmostOneTaskExecutor[%s]", task)
-                )),
+                )), ACL.SYSTEM),
```

Comment:
This fixes `Queue.maintain` to run as `SYSTEM`. Without that, `Queue.Task.getFullDisplayName()` is called when logging is enabled (and, as in #2791, even when it is not!) as `anonymous`, which can throw `AccessDeniedException` straight out of `maintain`¹, and…

¹Causing the queue to remove a `PendingItem` without ever adding the corresponding `BlockedItem`, so that `getItem(long)` suddenly and inexplicably starts returning null, breaking `ExecutorPickle`.